src/clims/models/extensible.py

```diff
@@ -0,0 +1,184 @@
+from __future__ import absolute_import, print_function
+
+from django.db import models
+from sentry.db.models import (Model, sane_repr)
+import json
+
+
+class ExtensibleType(Model):
+    """
+    Extensible objects are models that can be extended by plugin developers without
+    a django migration. They are defined programmatically in the plugin and registered
+    when running `lims upgrade`.
+
+    These types are examples of those that will be extensible, meaning that they can
+    be customized by plugins. Currently the Substance is the only one that's implemented.
+
+    * Substance
+    * Container
+    * Project
+    """
```

Comment:
I understand your concern and think it's great you're bringing it up early.

Note that substance, container and project will all have their own tables, because they'll have different requirements. We'll for example need ancestry information for `substances` but not `projects`. It's only the properties tables that will be reused for all of these. 

But this would still mean we'll have a huge properties table. That does not worry me because when we do queries in a DB we usually (preferably always for larger tables) really don't query the table under the hood, that's just a convenient abstraction. We are rather querying an index into the table. So in fact there will be different indexes for the `project` to `properties` relation than are for the `subject` to `properties` table. This way of approaching the problem moves most of the performance concerns from "domain developer" hands to the hands of people that know everything about DB peformance, in this case the postgres core team.

It also means that there will be a lot of extra data created (the indexes), but that will be an constant increase so it should not be a concern. (Also, data is cheap.)

I'm certain that with a moderately beefy DB server, this won't be an issue, even for millions of entries, but we'll write performance tests to show that later. Then we'll also address the lack of indexes.

Finally I'd like to add that I think we should never optimize prematurely, as long as there is a straightforward way to do it later, usually achieved by separation of concerns. If it would turn out I'm wrong, we would be able to fix things by only changing the data layer.

Reply:
Ok, thanks for the explanation, and interesting with the index thing. I also agree with the last thing, never to optimize prematurely.

Reply:
I want to mention the reason why I have concerns about performance. 1) clarity is very slow. I dont know why, udf system might contribute to this. 2) A general uncomfortness because I have never worked with this kind of solution before. I just have a diffuse feeling that many linked large tables may be slow. It would defenitely have some value if we could have a solution that is stress-tested in a production environment before.

Reply:
Yes. I think your concern is very valid, not in the least given our past experience with that system.

But as mentioned, I think it's improper implementation that causes this in most or all cases. Databases are extremely well optimized for this sort of thing and have been gradually improving since the 70s, but they are often misused. I look forward to doing some load tests on this. (I'd also like to mention that load testing on this without change would not show excellent performance since the indexes aren't all in place right now - I'm deferring that job to when the API isn't as much of a moving target.)